Test infrastructure, tests, CI, coverage and static analysis and more (closes #15)

blahah opened the pull request:
Large devops infrastructure setup PR

`Cakefile` manages most developer tasks, e.g.:
- `cake clean` removes generated JS files
- `cake build` regenerates JS from coffeescript and makes bin executable
- `cake test` runs tests including coverage analysis (and if on CI, pushes results to codecov.io)
- `cake test-nocov` as above but without coverage
- `cake docs` generates documentation from the comments

Travis-CI,  Codecov.io, and Code Climate are setup

MIT `LICENSE.md` and list of `CONTRIBUTORS.md`

blahah committed:
Setup test infrastructure

blahah committed:
Merge branch 'master' into tests

blahah committed:
extensions are in coffeescript

blahah committed:
separate out classes

blahah committed:
Making names more concise

blahah committed:
Add chai assertion lib

blahah committed:
Not JS!

blahah committed:
Coffee ignoring

blahah committed:
run tests via cake

blahah committed:
Bin is ignored

blahah committed:
Tweak ignoring build dirs

blahah committed:
Ignore debug log

blahah committed:
Test infrastructure set up

blahah committed:
Add istanbul coverage analysis

blahah committed:
now with docco docs

blahah committed:
deck api spec done

blahah committed:
Merge branch 'master' into tests

Conflicts:
	.gitignore
	Cakefile
	package.json

blahah committed:
license, ci, contributor list

blahah committed:
add community docs badge

blahah committed:
reorder

blahah committed:
add codecov

blahah committed:
Fix codecov

blahah committed:
Slide class specced, tested and implemented

blahah committed:
move codecov to npm script

blahah committed:
push codecov from travis

blahah commented:
From this point on tests should always be passing and coverage should only ever stay stable or increase :zap: :zap:

blahah commented:
last commit closes the following:
- #25 `slide.identifier()` gives a stable UUID
- #21 from JS object or from markdown with YAML frontmatter
- #23 and uuid is preserved with edits
- #22 `slide.clone()` gives a child with a new UUID, but all other properties preserved.

and contributes to the following:
- #30 slides have tags now
- #3 slides store custom styles and scripts (not yet implemented: they also need to be used in rendering)

blahah commented:
atomic commits henceforth :microscope: :space_invader:

TransGirlCodes commented:
Nice idea with the UUID - is there any danger of anyone editing these ID's - I guess if they are generated from the content of a slide itself, then there is no way?

blahah commented:
The UUIDs were originally generated from the slide content, but then I realise that broke the dag for parent-child relationships any time a slide is edited. So currently they are just random UUIDs.

I suppose we could go the git model and make slides immutable, so any edit changes the id by creating a new child?

blahah committed:
suppress log messages during tests

blahah committed:
collection specced and implemented

- collections have tags (closes #30)
- can fulltext / metadata search a collection (towards #24)
- load from / save to directory (reworked)
- names are optional with default
- add slides from markdown or as slide objects

blahah committed:
flushing

blahah committed:
use open access figures test data

blahah committed:
move oa figures to data dir

blahah committed:
More helpers

blahah committed:
collection-librarian refactor wip (squash me)
